configsuite/extension/ext.py

```diff
@@ -0,0 +1,206 @@
+    )
+
+    section_content += nodes.paragraph(text=schema.get(MK.Description, ""))
+
+    for config_key in [MK.ElementValidators, MK.ContextValidators]:
+        element_nodes = [
+            nodes.paragraph(text=validator.msg)
+            for validator in schema.get(config_key, [])
+        ]
+        section_content.extend(element_nodes)
+
+    for config_key in [MK.LayerTransformation, MK.Transformation]:
+        if config_key in schema:
+            section_content += nodes.paragraph(text=schema.get(config_key).msg)
+
+    section_children = cs_children(
+        classes=["cs_children"], ids=["cs_{}_children".format(title)]
+    )
+
+    if schema[MK.Type] == types.NamedDict:
+        example = dict()
+        for key, value in schema[MK.Content].items():
+            generate(value, section_children, key)
+
+        for child in section_children.children:
+            example[child["parent"]] = child["example"]
+
+    elif schema[MK.Type] == types.List:
+        generate(
+            schema[MK.Content][MK.Item], section_children, parent_title="List Item"
+        )
+        example = [section_children.children[0]["example"]]
+
+    elif schema[MK.Type] == types.Dict:
+        example = dict()
+        generate(schema[MK.Content][MK.Key], section_children, parent_title="Key")
+        generate(schema[MK.Content][MK.Value], section_children, parent_title="Value")
+        example[section_children.children[0]["example"]] = section_children.children[1][
+            "example"
+        ]
+
+    elif isinstance(schema[MK.Type], types.BasicType):
+        example = "{}".format(BASIC_EXAMPLE.get(schema[MK.Type].name, "undefined"))
+        section_children += nodes.paragraph(
+            text=":type: {_type}".format(_type=schema[MK.Type].name)
+        )
+
+    else:
+        err_msg = "Unexpected type ({}) in schema while generating documentation."
+        raise TypeError(err_msg.format(schema[MK.Type]))
+
+    section_content += section_children
+
+    section += section_content
+    section["example"] = example
+    section["parent"] = parent_title
+
+    parent_section += section
+
+
+class ConfigsuiteDocDirective(Directive):
+    has_content = True
+    option_spec = {"module": unchanged_required}
+
+    def run(self):
+        if "module" not in self.options:
+            raise self.error("Required :module: argument missing")
+
+        schema_function_string = self.options.get("module")
+
+        if "." not in schema_function_string:
+            raise self.error(
+                "Configsuite :module: has invalid format. "
+                + "Expected 'module.function', got {}".format(schema_function_string)
+            )
+
+        module_str, func_str = schema_function_string.rsplit(".", 1)
+
+        try:
+            module = importlib.import_module(module_str)
+        except (ImportError) as e:
+            raise self.error(str(e))
+
+        try:
+            schema = getattr(module, func_str)()
```

Comment:
The `func_str` is called in order to return the dict with a config_suite schema. Should this be implemented in a different manner? Should we validate that the dict is a valid schema?

Reply:
At least i would use module and func_str as argument to generate and return a schema. Not very found of funcs mutating input and no return

Reply:
> At least i would use module and func_str as argument to generate and return a schema. Not very found of funcs mutating input and no return.

Not sure if I follow you here. There is not modifying any arguments here. The `func_str` is just a string with the name of the function that we expect would return the `schema`. Given that the `func_str` indeed is a function name within the `module`, it's just called in order to retrieve the `schema`.

I wasn't sure if this was an okay method to use at all though, it seems a bit hacky, but then again how can it not? We're pretty much parsing a configuration file on the format of sphinx extensions.

> Not sure of the validation - i think just validating it's a dict and throw a proper error if not would be nice.

I'll do that!

Reply:
> Not sure if I follow you here. There is not modifying any arguments here. The `func_str` is just a string with the name of the function that we expect would return the `schema`. Given that the `func_str` indeed is a function name within the `module`, it's just called in order to retrieve the `schema`.
I misunderstood it seems the functionality behind `generate(...)`.. thought that one modified and generated the schema - which it doesn't .